cpp/src/arrow/dataset/dataset_writer.cc

```diff
@@ -549,11 +566,14 @@ class DatasetWriter::DatasetWriterImpl {
               WriteAndCheckBackpressure(std::move(batch), directory, prefix);
           if (!has_room.is_finished()) {
             // We don't have to worry about sequencing backpressure here since
-            // task_group_ serves as our sequencer.  If batches continue to arrive after
-            // we pause they will queue up in task_group_ until we free up and call
-            // Resume
+            // task_group_ serves as our sequencer.  If batches continue to arrive
+            // after we pause they will queue up in task_group_ until we free up and
+            // call Resume
             pause_callback_();
-            return has_room.Then([this] { resume_callback_(); });
+            paused_ = true;
+            return has_room.Then([this] { ResumeIfNeeded(); });
+          } else {
+            ResumeIfNeeded();
           }
```

Comment:
So this is because when `has_room` is finished when `paused_`, `resume_callback_` is not called?

Reply:
If `has_room` is finished then we can unpause if there are no other tasks because there is room for another batch.

Reply:
Usally mutex is used with mutable. But this LGTM either

Reply:
May I ask a stupid question that why `has_room.Then` not being called in this scenerio?

Reply:
Oh because it's paused...I got to understand this. So we don't "resume" enough, which causing new tasks didn't being consumed.

Reply:
Here is a picture.  In the old behavior, we resume too much.
![Dataset Backpressure](https://github.com/apache/arrow/assets/1696093/1a330f4c-0d00-4ac4-abf7-944d86309aff)